### PR TITLE
Improve search activity

### PR DESCRIPTION
**Pull request by MrChip53**
**Changes**
- [x] Use CustomListRowPresenter in SearchProvider.java
- [x] Fix background in search activity
- [ ] ~~Add rating to search cards?~~ no, no and no

Requires #703 

**Issues**

#702

### CODE REVIEW

**Comment by nielsvanvelzen** [comment]
Marking this one as draft until #703 is merged

**Comment by nielsvanvelzen** [comment]
@MrChip53 can you rebase this PR and mark it ready for review again?

File: app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt [hunk 1]
```diff
@@ -174,6 +175,48 @@ class BackgroundService(
...
+
+		//Manually grab the backdrop URL
+		val options = ImageOptions()
+		options.maxWidth = 1200
+		options.imageType = ImageType.Backdrop
+		options.tag = searchHint.backdropImageTag
+
+		val backdropUrl = apiClient.GetImageUrl(searchHint.backdropImageItemId, options)
+
+		if (backdropUrl.isNullOrEmpty()) return clearBackgrounds()
```
Comment (thornbill) [inline app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt:194]: It seems like all of the code from this point on is common with the other `setBackground` ~method~ function. It would probably be best to create a private function for this that just accepts the url string.

**Review by thornbill — CHANGES_REQUESTED** [review]

**Review by thornbill — CHANGES_REQUESTED** [review]
This is crashing for me when I try to change to a different row in the search activity. It may be specific to switching to a Person item? Stack trace is below:

```
2021-02-01 12:34:13.347 6003-6003/org.jellyfin.androidtv.debug E/MessageQueue-JNI: java.lang.IllegalArgumentException: itemId
        at org.jellyfin.apiclient.interaction.BaseApiClient.GetImageUrl(BaseApiClient.java:767)
        at org.jellyfin.androidtv.data.service.BackgroundService.setBackground(BackgroundService.kt:166)
        at org.jellyfin.androidtv.ui.search.LeanbackSearchFragment.lambda$onCreate$1$LeanbackSearchFragment(LeanbackSearchFragment.java:46)
        at org.jellyfin.androidtv.ui.search.-$$Lambda$LeanbackSearchFragment$kyLRXuEYxkFa8XhsK2W9ivzAoao.onItemSelected(Unknown Source:4)
        at org.jellyfin.androidtv.ui.browsing.CompositeSelectedListener.onItemSelected(CompositeSelectedListener.java:26)
        at org.jellyfin.androidtv.ui.browsing.CompositeSelectedListener.onItemSelected(CompositeSelectedListener.java:11)
        at androidx.leanback.app.SearchSupportFragment$7.onItemSelected(SearchSupportFragment.java:351)
        at androidx.leanback.app.SearchSupportFragment$7.onItemSelected(SearchSupportFragment.java:341)
        at androidx.leanback.widget.ListRowPresenter.dispatchItemSelectedListener(ListRowPresenter.java:597)
        at androidx.leanback.widget.RowPresenter.onRowViewSelected(RowPresenter.java:512)
        at androidx.leanback.widget.ListRowPresenter.onRowViewSelected(ListRowPresenter.java:605)
        at androidx.leanback.widget.RowPresenter.setRowViewSelected(RowPresenter.java:434)
        at androidx.leanback.app.RowsSupportFragment.setRowViewSelected(RowsSupportFragment.java:322)
        at androidx.leanback.app.RowsSupportFragment.onRowSelected(RowsSupportFragment.java:252)
        at androidx.leanback.app.BaseRowSupportFragment$1.onChildViewHolderSelected(BaseRowSupportFragment.java:55)
        at androidx.leanback.widget.GridLayoutManager.fireOnChildViewHolderSelected(GridLayoutManager.java:926)
        at androidx.leanback.widget.GridLayoutManager.dispatchChildSelected(GridLayoutManager.java:1013)
        at androidx.leanback.widget.GridLayoutManager.scrollToView(GridLayoutManager.java:2966)
        at androidx.leanback.widget.GridLayoutManager.scrollToView(GridLayoutManager.java:2949)
        at androidx.leanback.widget.GridLayoutManager.onRequestChildFocus(GridLayoutManager.java:2887)
        at androidx.recyclerview.widget.RecyclerView$LayoutManager.onRequestChildFocus(RecyclerView.java:10318)
        at androidx.recyclerview.widget.RecyclerView.requestChildFocus(RecyclerView.java:3026)
        at android.view.ViewGroup.requestChildFocus(ViewGroup.java:819)
        at android.view.ViewGroup.requestChildFocus(ViewGroup.java:819)
        at android.view.ViewGroup.requestChildFocus(ViewGroup.java:819)
        at androidx.recyclerview.widget.RecyclerView.requestChildFocus(RecyclerView.java:3029)
        at android.view.ViewGroup.requestChildFocus(ViewGroup.java:819)
        at android.view.View.handleFocusGainInternal(View.java:6926)
        at android.view.ViewGroup.handleFocusGainInternal(ViewGroup.java:795)
        at android.view.View.requestFocusNoSearch(View.java:11546)
        at android.view.View.requestFocus(View.java:11520)
        at android.view.ViewGroup.requestFocus(ViewGroup.java:3230)
        at android.view.ViewRootImpl$ViewPostImeInputStage.performFocusNavigation(ViewRootImpl.java:4976)
        at android.view.ViewRootImpl$ViewPostImeInputStage.processKeyEvent(ViewRootImpl.java:5093)
        at android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:4905)
        at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:4426)
        at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:4479)
        at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:4445)
        at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:4585)
        at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:4453)
        at android.view.ViewRootImpl$AsyncInputStage.apply(ViewRootImpl.java:4642)
        at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:4426)
        at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:4479)
        at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:4445)
        at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:4453)
        at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:4426)
    	at android.view.ViewRootImpl$InputStage.onDeliverToN
```

**Comment by nielsvanvelzen** [comment]
Got the same error as @thornbill 
source of `GetImageUrl` for reference:
```java
	/** 
	 Gets an image url that can be used to download an image from the api
	 
	 @param itemId The Id of the item
	 @param options The options.
	 @return System.String.
	 @exception System.ArgumentNullException itemId
	*/
	public final String GetImageUrl(String itemId, ImageOptions options)
	{
		if (tangible.DotNetToJavaStringHelper.isNullOrEmpty(itemId))
		{
			throw new IllegalArgumentException("itemId");
		}

		String url = "Items/" + itemId + "/Images/" + options.getImageType();

		return GetImageUrl(url, options, new QueryStringDictionary());
	}
```

File: app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt [hunk 2]
```diff
@@ -145,23 +146,46 @@ class BackgroundService(
...
+		if (searchHint == null || !userPreferences[UserPreferences.backdropEnabled])
+			return clearBackgrounds()
+
+		//Manually grab the backdrop URL
+		val options = ImageOptions()
+		options.maxWidth = 1200
+		options.imageType = ImageType.Backdrop
+		options.tag = searchHint.backdropImageTag
+
+		val backdropUrls = setOf(apiClient.GetImageUrl(searchHint.backdropImageItemId, options))
```
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt:166]:
```suggestion
		val backdropUrls = setOfNotNull(searchHint.backdropImageItemId?.let { apiClient.GetImageUrl(it, options) })
```

File: app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt [hunk 3]
```diff
@@ -145,23 +146,46 @@ class BackgroundService(
...
+	 * Use one available backdrop from [SearchHint] as background.
+	 */
+	fun setBackground(searchHint: SearchHint?) {
+		// Check if item is set and backgrounds are enabled
+		if (searchHint == null || !userPreferences[UserPreferences.backdropEnabled])
+			return clearBackgrounds()
+
+		//Manually grab the backdrop URL
+		val options = ImageOptions()
+		options.maxWidth = 1200
```
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt:162]:
```suggestion
```
No need for this

File: app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt [hunk 4]
```diff
@@ -145,23 +146,46 @@ class BackgroundService(
...
+						Glide.with(context)
+								.load(url)
+								.override(windowSize.width, windowSize.height)
+								.centerCrop()
+								.submit()
+					}
+					.map { future -> async { future.get() } }
+					.awaitAll()
+					.onEach { it.colorFilter = colorFilter }
+					.filterNotNull()
```
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt:188]: Why did you change the indentation?

File: app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt [hunk 5]
```diff
@@ -145,23 +146,46 @@ class BackgroundService(
...
+
+	/**
+	 * Use one available backdrop from [SearchHint] as background.
+	 */
+	fun setBackground(searchHint: SearchHint?) {
+		// Check if item is set and backgrounds are enabled
+		if (searchHint == null || !userPreferences[UserPreferences.backdropEnabled])
+			return clearBackgrounds()
+
+		//Manually grab the backdrop URL
```
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt:160]:
```suggestion
		// Manually grab the backdrop URL
```

File: app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt [hunk 6]
```diff
@@ -145,23 +146,46 @@ class BackgroundService(
...
+		loadBackgrounds(backdropUrls)
+	}
+
+	/**
+	 * Use one available backdrop from [SearchHint] as background.
+	 */
+	fun setBackground(searchHint: SearchHint?) {
+		// Check if item is set and backgrounds are enabled
+		if (searchHint == null || !userPreferences[UserPreferences.backdropEnabled])
+			return clearBackgrounds()
```
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt:158]: Move this part (from both functions) to the `loadBackgrounds` function to prevent duplicate code.

File: app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt [hunk 7]
```diff
@@ -145,23 +146,46 @@ class BackgroundService(
 		val parentBackdropUrls = baseItem.parentBackdropImageTags.getUrls(baseItem.parentBackdropItemId)
 		val backdropUrls = itemBackdropUrls.union(parentBackdropUrls)
 
+		loadBackgrounds(backdropUrls)
+	}
+
+	/**
+	 * Use one available backdrop from [SearchHint] as background.
```
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt:153]:
```suggestion
	 * Use backdrop from [SearchHint] as background.
```

**Review by nielsvanvelzen — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt (hunk 2 above)
Comment (nielsvanvelzen) [reply app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt:166]: This one should fix the crash

File: app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt (hunk 6 above)
Comment (MrChip53) [reply app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt:158]: I need some more explanation on this. The load backgrounds function is supposed to be generic and not be aware of whether we have a BaseItem or SearchHint type. This change would make the loadBackground function aware. If this change is really wanted I would suggest maybe just moving everything back into setBackground with a any param and check if that param is not null and is searchhint or baseitem.

`/**
	 * Use all available backdrops from [baseItem] or [SearchHint] as background.
	 */
	fun setBackground(baseItem: Any?) {
		// Check if item is set and backgrounds are enabled
		if (baseItem == null || !userPreferences[UserPreferences.backdropEnabled])
			return clearBackgrounds()

		val backdropUrls: Set<String>

		when (baseItem) {
			is BaseItemDto -> {
				// Get all backdrop urls
				val itemBackdropUrls = baseItem.backdropImageTags.getUrls(baseItem.id)
				val parentBackdropUrls = baseItem.parentBackdropImageTags.getUrls(baseItem.parentBackdropItemId)
				backdropUrls = itemBackdropUrls.union(parentBackdropUrls)
			}
			is SearchHint -> {
				val options = ImageOptions()
				options.imageType = ImageType.Backdrop
				options.tag = baseItem.backdropImageTag

				backdropUrls = setOfNotNull(baseItem.backdropImageItemId?.let { apiClient.GetImageUrl(it, options) })
			}
			else -> {
				return clearBackgrounds();
			}
		}

		loadBackgrounds(backdropUrls)
	}`

something like this maybe?

Comment (nielsvanvelzen) [reply app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt:158]: Ignore my message, the parameter is different in both functions so the code is not duplicated. I was only looking at the preferences part.

File: app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt [hunk 8]
```diff
@@ -145,6 +146,28 @@ class BackgroundService(
 		val parentBackdropUrls = baseItem.parentBackdropImageTags.getUrls(baseItem.parentBackdropItemId)
 		val backdropUrls = itemBackdropUrls.union(parentBackdropUrls)
 
+		loadBackgrounds(backdropUrls)
+	}
+
+	/**
+	 * Use backdrop from [SearchHint] as background.
+	 */
+	fun setBackground(searchHint: SearchHint?) {
```
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt:155]: Does this actually need to be nullable? All items in the search results are searchhints so they can't be null.

Comment (MrChip53) [reply app/src/main/java/org/jellyfin/androidtv/data/service/BackgroundService.kt:155]: Sure doesn't lol

**Review by nielsvanvelzen — APPROVED** [review]

**Review by thornbill — APPROVED** [review]
I noticed a few item types aren't being displayed in the correct aspect ratio (programs, albums, and songs seem to be using the poster ratio). This can be addressed later though.
